File: LeetCode/322. Coin Change/coin_change.py
# DP 问题
# dp[i] = 1 if i == coin for coin in coins
# dp[i] = min(dp[i-coin]) + 1 for coin in coins and i > coin
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [-1 for _ in range(amount + 1)]
        dp[0] = 0
        
        for i in range(1, amount + 1):
            for coin in coins:
                if i == coin:
                    dp[i] = 1
                elif i > coin:
                    if dp[i-coin] != -1:
                        if dp[i] == -1:
                            dp[i] = dp[i-coin] + 1 
                        else:
                            dp[i] = min(dp[i], dp[i-coin] + 1) 
        return dp[-1]

File: LeetCode/322. Coin Change/coin_change.cpp
//完全背包问题
//变化是正好要求背包装满（正好凑齐amount）

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        if(amount==0) return 0;
        int n = coins.size();
        int dp[amount+1];
        
        //初始化dp数组，因为要求最后要正好装满背包，求的又是最小问题(需要最少的硬币)
        //所以除了dp[0]其他都赋值为最大值
        //这里用amount+1就行了
        for(int j=1;j<=amount;j++){
            dp[j] = amount+1;
        }
        dp[0] = 0;
        
        for(int i=0;i<n;++i){
            for(int j=coins[i];j<=amount;++j){
                dp[j] = min(dp[j], dp[j-coins[i]]+1);
            }
        }
        
        if(dp[amount] > amount) return -1;
        else return dp[amount];
        
    }
};
